cpp/src/arrow/compute/kernels/vector_run_length_encode.cc

```diff
@@ -0,0 +1,393 @@
+    this->input_validity = this->input_array.buffers[0].data;
+    this->input_values = this->input_array.buffers[1].data;
+    int64_t input_offset = this->input_array.offset;
+
+    this->read_offset = input_offset;
+    Element element = this->ReadValue();
+    int64_t num_values_output = 1;
+
+    // calculate input null count by ourselves. The input span likely got sliced by an
+    // ExecSpanIterator using SetOffset right before this code executes. SetOffset leaves
+    // the null_count value of the ArraySpan as kUnknownNullCount.
+    int64_t input_null_count = element.valid ? 0 : 1;
+    int64_t output_null_count = input_null_count;
+    for (int64_t input_position = 1; input_position < this->input_array.length;
+         input_position++) {
+      this->read_offset = input_offset + input_position;
+      Element previous_element = element;
+      element = this->ReadValue();
+      if (element != previous_element) {
+        num_values_output++;
+        if (!element.valid) {
+          output_null_count++;
+        }
+      }
+      if (!element.valid) {
+        input_null_count++;
+      }
+    }
+
+    std::shared_ptr<Buffer> validity_buffer = NULLPTR;
+    // in bytes
+    int64_t validity_buffer_size = 0;
+    if (has_validity_buffer) {
+      validity_buffer_size = bit_util::BytesForBits(num_values_output);
+      ARROW_ASSIGN_OR_RAISE(
+          validity_buffer,
+          AllocateBuffer(validity_buffer_size, this->kernel_context->memory_pool()));
+    }
+    ARROW_ASSIGN_OR_RAISE(auto values_buffer,
+                          AllocateBuffer(bit_util::BytesForBits(num_values_output *
+                                                                ArrowType().bit_width()),
+                                         this->kernel_context->memory_pool()));
+    ARROW_ASSIGN_OR_RAISE(auto run_lengths_buffer,
+                          AllocateBuffer(num_values_output * sizeof(int32_t),
+                                         this->kernel_context->memory_pool()));
+
+    output_array_data->length = this->input_array.length;
+    output_array_data->offset = 0;
+    output_array_data->buffers = {NULLPTR};
+    auto values_array_data =
+        ArrayData::Make(this->input_array.type->GetSharedPtr(), num_values_output);
+    auto run_ends_array_data = ArrayData::Make(int32(), num_values_output);
+    values_array_data->buffers.push_back(std::move(validity_buffer));
+    values_array_data->buffers.push_back(std::move(values_buffer));
+    run_ends_array_data->buffers.push_back(NULLPTR);
+    run_ends_array_data->buffers.push_back(std::move(run_lengths_buffer));
+
+    output_array_data->null_count.store(input_null_count);
+    values_array_data->null_count = output_null_count;
+
+    this->output_validity = values_array_data->template GetMutableValues<uint8_t>(0);
+    this->output_values = values_array_data->template GetMutableValues<uint8_t>(1);
+    auto output_run_ends = run_ends_array_data->template GetMutableValues<int32_t>(1);
```

Comment:
```suggestion
    // set mutable pointers for output; `WriteValue` uses member `output_` variables
    this->output_validity = values_array_data->template GetMutableValues<uint8_t>(0);
    this->output_values = values_array_data->template GetMutableValues<uint8_t>(1);
    auto output_run_ends = run_ends_array_data->template GetMutableValues<int32_t>(1);
```

I think it would be helpful to add a comment so that it's clearer to a reader that these member variables are aliases and are used by WriteValue. It can be hard to follow the logic otherwise.